### ISSUE TITLE
Understanding list api calls

### ISSUE THREAD

**Issue by saswatac** [Azure/azure-storage-fuse#646]
### Which version of the blobfuse was used?
1.3.8

### Which OS (please include version) are you using?
```
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

### What problem was encountered?

80-90% of the blob storage costs is coming from list blob api calls. As per suggestions in issue https://github.com/Azure/azure-storage-fuse/issues/596, after enabling caching we had seen temporary reduction, but as 
the container sizes have grown, the cost has shot back up.

I have observed that for each list, blobfuse calls listblob api at each node in the directory tree.
For example, for container mounted at `/mount` the command `ls /mount/datasets/bike/raw` results in 4 list calls - 

```
Sep 16 08:49:33 matterhorn blobfuse[15536]:  ==> REQUEST/RESPONSE :: GET https://ncsazuretest.blob.core.windows.net/neuralconcept-ncstest?comp=list&delimiter=/&include=metadata&maxresults=5000&prefix=datasets&restype=container?&User-Agent=azure-storage-fuse/1.3.8&x-ms-date=Thu, 16 Sep 2021 08:49:33 GMT&x-ms-version=2018-11-09&Authorization=****&Transfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 200 :: REQ ID : 2fc040f5-a01e-0015-63d7-aa9257000000
Sep 16 08:49:33 matterhorn blobfuse[15536]:  ==> REQUEST/RESPONSE :: GET https://ncsazuretest.blob.core.windows.net/neuralconcept-ncstest?comp=list&delimiter=/&include=metadata&maxresults=5000&prefix=datasets/bike&restype=container?&User-Agent=azure-storage-fuse/1.3.8&x-ms-date=Thu, 16 Sep 2021 08:49:33 GMT&x-ms-version=2018-11-09&Authorization=****&Transfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 200 :: REQ ID : 33c3201f-201e-008c-4bd7-aa1d95000000
Sep 16 08:49:33 matterhorn blobfuse[15536]:  ==> REQUEST/RESPONSE :: GET https://ncsazuretest.blob.core.windows.net/neuralconcept-ncstest?comp=list&delimiter=/&include=metadata&maxresults=5000&prefix=datasets/bike/raw&restype=container?&User-Agent=azure-storage-fuse/1.3.8&x-ms-date=Thu, 16 Sep 2021 08:49:33 GMT&x-ms-version=2018-11-09&Authorization=****&Transfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 200 :: REQ ID : 27723f53-201e-00c3-19d7-aad98d000000
Sep 16 08:49:33 matterhorn blobfuse[15536]:  ==> REQUEST/RESPONSE :: GET https://ncsazuretest.blob.core.windows.net/neuralconcept-ncstest?comp=list&delimiter=/&include=metadata&maxresults=5000&prefix=datasets/bike/raw/&restype=container?&User-Agent=azure-storage-fuse/1.3.8&x-ms-date=Thu, 16 Sep 2021 08:49:33 GMT&x-ms-version=2018-11-09&Authorization=****&Transfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 200 :: REQ ID : fc32a4eb-b01e-00ab-30d7-aa87dc000000
```

Is there a way to avoid all these calls?

I have also noticed that if i call `ls /mount/datasets/bike/raw` within the `attr_timeout`, then it makes only the last or last two calls, but makes all the 4 above calls if called after  `attr_timeout` even though `use-attr-cache=true` . Is this expected?

The calls have a pagination limit of 5000, so i am afraid that for large containers, it would result in lots of calls.

### If relevant, please share your mount command.

```
blobfuse $MOUNT_DIR --container-name=$BUCKET --tmp-path=/mnt/data/blobfusetmp/ -o allow_other --use-attr-cache=true -o attr_timeout=240 -o entry_timeout=240 -o negative_timeout=240 --basic-remount-check=true --log-level=LOG_DEBUG
```

**Comment by vibhansa-msft** [comment]
Putting my answers to each query mentioned above :: 

1. Pagination limit of 5000 : this limit is from azure-storage itself and not a blobfuse induced restriction. If you have more then 5000 items in the path there will be multiple trip to retrieve entire list.

2. attr-cache :  it only caches the attributes of files and not the directory listing. So executing a 'ls' will always result in list-api call.

3. multiple list-api call : When a 'ls' call is executed linux first tries to get the attributes of the directory and then do a list call for that. This will happen at each path in hierarchy so if its 'ls A/B/C' then it will be repeated for all 'A', 'A/B', and 'A/B/C'. This is as per linux kernel functionality and nothing special done by blobfuse. If attribute cache is not there then even get-attribute will result into a list call. Once you enable attr-cache those calls are saved. However list calls made by system will still be three 'A', 'A/B', 'A/B/C' and each will result into a call to container. If there are more then 5K items in a path then there will be multiple calls for that path until entire list is retrieved. There is no way to save this from kernel or blobfuse end. Only way to reduce this is to not fire those commands unless required.

- As log_debug are enable, you can share the log file and I can point you to logs confirming behavior in (3)

**Comment by vibhansa-msft** [comment]
For the attr_timeout part : this specifies how long will kernel cache the attributes. So any call to get-attribute is done in this time window, kernel will serve it from its own cache and call will not land in blobfuse.

**Comment by vibhansa-msft** [comment]
(Ignore previous reply on this thread, updating some facts here)

Putting my answers to each query mentioned above ::

1. Pagination limit of 5000 : this limit is from azure-storage itself and not a blobfuse induced restriction. If you have more then 5000 items in the path there will be multiple trip to retrieve entire list.

2. attr-cache : it only caches the attributes of files and not the directory listing. So executing a 'ls' will always result in list-api call.

3. multiple list-api call : When a 'ls' call is executed linux first tries to get the attributes of the directory and then do a list call for that. get-attribute  will happen at each path in hierarchy so if its 'ls A/B/C' then it will be repeated for all 'A', 'A/B', and 'A/B/C'.  'attr-timeout' that you have configured configures how long kernel can cache the attributes. So any call made within this limit for get-attribute, can be served by the kernel itself. When configured timeout elapse, get-attribute will be forwarded to blobfuse. 'attr-cache' enables caching of attributes in blobfuse but that's only for file and not for directories. Hence get-attribute call for a directory will result into list call on storage. And this will get 5K elements in one call and continue till all elements are retrieved.

As log_debug are enable, you can share the log file and I can point you to logs confirming behavior in (3)

**Comment by vibhansa-msft** [comment]
In log_debug we can observe there were multiple get-attributes calls for each element in the given path and then at the end it was a read-dir call.

**Comment by saswatac** [comment]
Thanks for your response. It makes things clearer.

Regarding 3, 

> When configured timeout elapse, get-attribute will be forwarded to blobfuse.

Lets assume this structure -

```
└── A
    ├── B
    │   └── C
    └── data
```

and lets say data has 1 million files.

if i am doing `ls A/B/C` and cache has expired. To get attribute for "A", a list call with prefix="A" is done. Will this mean that this call will iterate through all 1 million files under data ?

**Comment by vibhansa-msft** [comment]
No if you have done A/B/C we will do list of A and A/B only to get attribute and hence only one list-api call will be done for them. However if A/B/C has 1m objects then that will end up in having 100s of iterations to get entire list. You can confirm this with the logs that you shared earlier.

Sep 16 08:49:33 matterhorn blobfuse[15536]:  ==> REQUEST/RESPONSE :: GET https://ncsazuretest.blob.core.windows.net/neuralconcept-ncstest?comp=list&delimiter=/&include=metadata&maxresults=5000&prefix=datasets&restype=container?&User-Agent=azure-storage-fuse/1.3.8&x-ms-date=Thu, 16 Sep 2021 08:49:33 GMT&x-ms-version=2018-11-09&Authorization=****&Transfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 200 :: REQ ID : 2fc040f5-a01e-0015-63d7-aa9257000000


In this log look for "prefix=datasets" string, this means its doing list of "datasets" directory, while "prefix=datasets/bike/raw/" means listing of datasets/bike/raw path.

**Comment by saswatac** [comment]
Thanks for the clarification. I also confirm i see this behaviour after enabling debug logs on one of our prod containers. 

One small suggestion would be then to use maxresults=1  for the list api call which is done to get the attribute ( assuming 1 is enough, i see that maxresults=1 is used for the initial list call when blob is mounted)

**Comment by vibhansa-msft** [comment]
When container is mounted we just want to ensure that the authentication part is going through fine and hence we just try to list something. 
While doing attribute retrieval, we used to do maxResults=2 earlier but there is an issue with that. Store returns the list in lexicographical order and when you want to get attributes of a directory we do list with prefix=<dir path>. This prefix matches with all the files/directories having same name. In block blob account even the children of this directory will be part of list. In one customer scenario we observed they have mutiple files/directories with similar name and this result '2' was not having the directory we were searching. Hence as a fix we moved to a list of 5000 and go even further down for next calls if the intended directory is not in the list.
